### PR TITLE
Allow registering specific methods to use as constructors

### PR DESCRIPTION
**Pull request by matt-gardner**
Fixes one of the two main problems in #3389. This makes it so that you can specify a particular method to use as a constructor when registering a class, removing one of the main cases where we needed to write a custom `from_params` method.  Those methods are not self-documenting, as you don't know what's inside the `params` argument, and they make it very hard to use our code without also buying into our config file architecture.

This PR isn't actually as big as it looks.  I took the opportunity to re-order some of the methods in `Vocabulary` (grouping together the various constructors), which is the bulk of the actual diff here.

There are two main classes that were the target for this change: `Vocabulary` and `Embedding`.  `Vocabulary` has three different constructors, depending on how you want to create the vocab, and so is registered three times.  That's the place to look to see the benefit of this change.

The place to look to see the main logic that changed is in `FromParams`, and a little bit in `Registrable`.  Everything else is either using the new functionality (`Vocabulary` and `Embedding`) or minor things to account for how calling the classes changed (and this is mostly in tests, but also in a few other places).

A side effect of how I did this was that I needed to add `**extras` to every remaining custom `from_params` method.  The plan is to remove those entirely, eventually, so I didn't worry about it too much.  Some of them will need a bit of redesign of how things are constructed and passed around, though (mostly in the callbacks).

### CODE REVIEW

**Comment by DeNeutoy** [comment]
@matt-gardner sorry this caught me just as I was leaving for Christmas. Is this in a reviewable state or...? Looks like quite a few conflicts.

**Comment by matt-gardner** [comment]
Hmm, I thought it was in a reviewable state; let me look at it again.  I'll let you know when it's ready.

**Comment by matt-gardner** [comment]
If there are still failure at this point, they are minor and shouldn't affect a code review.  So I'd say it's good to look at now.

File: allennlp/common/registrable.py [hunk 1]
```diff
@@ -38,18 +38,22 @@ class Registrable(FromParams):
...
 
-    _registry: Dict[Type, Dict[str, Type]] = defaultdict(dict)
+    _registry: Dict[Type, Dict[str, Tuple[Type, str]]] = defaultdict(dict)
     default_implementation: str = None
 
     @classmethod
-    def register(cls: Type[T], name: str, exist_ok=False):
+    def register(cls: Type[T], name: str, constructor: str = None, exist_ok: bool = False):
         """
         Register a class under a particular name.
```
Comment (DeNeutoy) [inline allennlp/common/registrable.py]: Now we can have markdown in docstrings, can you add a couple of examples of registering different constructors and their corresponding jsonnet config?

**Comment by brendan-ai2** [comment]
@matt-gardner, looks like this caused some issues in `allennlp-hub`. See http://build.allennlp.org/viewLog.html?buildId=24152&buildTypeId=AllenNLPHub_Master&tab=buildLog&branch_AllenNLPHub=%3Cdefault%3E.

**Comment by matt-gardner** [comment]
Can we get a notification for these breakages in our slack channel?

**Comment by brendan-ai2** [comment]
Absolutely. I'll figure out how to do that.

**Comment by matt-gardner** [comment]
Also, FYI, the tests are broken because I removed the old, deprecated way of passing arguments to a `BasicTextFieldEmbedder` (part of my effort to squash custom `from_params` methods).  We just need to modify the config files inside all of those `model.tar.gz` files, to include a `token_embedders` wrapper around their embedder arguments.  I'll add a note to the breaking changes in the release notes.  I'm not in the best position to fix that issue at the moment, as I'm on a slow internet connection at home, and downloading and re-uploading those models would be a pain.  I'll open an issue on the hub repo to fix this, and hopefully get to it on Friday when I'm in the office.

**Comment by brendan-ai2** [comment]
https://github.com/allenai/allennlp-hub/issues/13

**Pull request merged**